File: C++/majority-element.cpp
#include <vector>


class Solution {
public:
    int majorityElement(std::vector<int>& nums) {
        // Time:  O(n)
        // Space: O(1)
        int element = nums[0];
        int counter = 0;
        for (int num: nums) {
            if (counter == 0) {
                element = num;
                counter++;
            } else {
                if (num == element) {
                    counter++;
                } else {
                    counter--;
                }
            }
        }
        return element;
    }
};

File: Python/majority-element.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # Time:  O(n)
        # Space: O(1)
        element = None
        counter = 0
        for num in nums:
            if counter == 0:
                element = num
                counter += 1
            else:
                if num == element:
                    counter += 1
                else:
                    counter -= 1
        return element
